cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,163 @@
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> InboundUpgrade<C> for InboundProtocolConfig
+where
+    C: AsyncRead + Unpin + Send + 'static,
+{
+    type Output = Message;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_inbound(self, mut socket: C, info: Self::Info) -> Self::Future {
+        trace!(
+            "Upgrading inbound connection: {}",
+            String::from_utf8_lossy(info)
+        );
+        Box::pin(async move {
+            let message = upgrade::read_one(&mut socket, 1024).await?;
+            let mut de = serde_json::Deserializer::from_slice(&message);
+            let info = Message::deserialize(&mut de)?;
+
+            Ok(info)
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("read: ")]
+    Read(#[from] upgrade::ReadOneError),
+    #[error("deserialize: ")]
+    Deserialize(#[from] serde_json::Error),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::prelude::*;
+    use libp2p::core::{
+        multiaddr::multiaddr,
+        transport::{memory::MemoryTransport, ListenerEvent, Transport},
+        upgrade,
+    };
+    use rand::{thread_rng, Rng};
+    use std::str::FromStr;
+
+    #[tokio::test]
+    async fn correct_transfer() {
+        let mem_addr = multiaddr![Memory(thread_rng().gen::<u64>())];
+        let mut listener = MemoryTransport.listen_on(mem_addr).unwrap();
+
+        let send_msg = Message {
+            swap_id: SwapId::from_str("ad2652ca-ecf2-4cc6-b35c-b4351ac28a34").unwrap(),
+            secret_hash: SecretHash::from_str(
+                "bfbfbfbfbfbfbfbfbfbfbfbfbfbfbfbf\
+                 bfbfbfbfbfbfbfbfbfbfbfbfbfbfbfbf",
+            )
+            .unwrap(),
+        };
+
+        let listener_addr =
+            if let Some(Some(Ok(ListenerEvent::NewAddress(a)))) = listener.next().now_or_never() {
+                a
+            } else {
+                panic!("MemoryTransport not listening on an address!");
+            };
+
+        tokio::task::spawn({
+            async move {
+                let listener_event = listener.next().await.unwrap();
+                let (listener_upgrade, _) = listener_event.unwrap().into_upgrade().unwrap();
+                let conn = listener_upgrade.await.unwrap();
+
+                let config = OutboundProtocolConfig { msg: send_msg };
+                upgrade::apply_outbound(conn, config, upgrade::Version::V1)
+                    .await
+                    .unwrap();
+            }
+        });
+
+        let conn = MemoryTransport.dial(listener_addr).unwrap().await.unwrap();
+
+        let config = InboundProtocolConfig {};
+        let msg = upgrade::apply_inbound(conn, config).await.unwrap();
+
+        assert_eq!(send_msg, msg)
```

Comment:
I think the order is "actual", "expected".

```suggestion
        assert_eq!(received_msg, sent_msg)
```